### ISSUE TITLE
Why we need remove and add self loop in GAT?

### ISSUE THREAD

**Issue by mrbeann** [rusty1s/pytorch_geometric#138]
## ❓ Questions & Help

When I read the code in nn.conv.gat_conv.py, I found it firstly remove the loop and then add it. Why this is necessary?
```python
    def forward(self, x, edge_index):
        """"""
        edge_index, _ = remove_self_loops(edge_index)
        edge_index = add_self_loops(edge_index, num_nodes=x.size(0))
```

**Comment by rusty1s** [comment]
We do this in case there are already (a few) self-loops in the graph. Normally, `remove_self_loops` is a no-op.

**Comment by mrbeann** [comment]
Thanks for your reply!

**Comment by RuihongQiu** [comment]
Hi, what if there is `edge_attr` attribute in `data`? After applying `remove_self_loops`, which also removes the `edge_attr`, how could to the weighted self loop?

**Comment by rusty1s** [comment]
I‘m not sure I understand. The GAT conv does not take in edge weights. PyG Lperators that take in edge attributes do not add self-loops and instead provide a skip connection via the `root_weight` flag. Users can then decide to manually add self-loops or use the skip connection.

**Comment by RuihongQiu** [comment]
Sorry for the confuse.

Let's assume that we have a directed weighted graph with some of the nodes containing self-loop. If we gonna design a self-attention for convolution, we need to add a weighted self-loop to those nodes without self-loop.

I think that nowthat the `remove_self_loops` can deal with the `edge_attr`, if there is any chance to let `add_self_loop` have the same feature.

**Comment by rusty1s** [comment]
This could only be achieved if we limit edge attributes to be a one-dimensional weight vector. It sounds reasonable to integrate it into `add_self_loop`.

**Comment by RuihongQiu** [comment]
Yes, it should only be simple weighted edge. If not this case, there is no reasonable way to initial the edge weight. For simple weighted edge, a self loop can be assigned as weight `1`.

I think that the origin self loop containing weight, should not be removed as well. Do I make sense?

**Comment by rusty1s** [comment]
Added in master.

**Comment by RuihongQiu** [comment]
I saw the update! Cool!

One small concern though: what if there is a self loop with weight other than `1`, say `2`, which is removed by `remove_self_loop`? I think that weight still contains extra information. Seems like it should be added back by `add_self_loop` of not be removed by `remove_self_loop`?

**Comment by rusty1s** [comment]
I understand your concern, but this is something which can not be efficiently implemented. In the end, this results in a slow `isin` call.

**Comment by RuihongQiu** [comment]
I agree. It needs to check every node if it is a pre-assigned self loop. Slow operation.

**Comment by rusty1s** [comment]
The best would be to add a flag to the GCN operator which determines if the operator itself should add the self-loops or the user.

**Comment by rusty1s** [comment]
Wait, this can be efficiently implemented, *e.g.*:
```python
mask = row == col
masked_weight = edge_weight[mask]
edge_index, edge_weight = remove_self_loops(...)
loop_weight = torch.full((num_nodes, ), fill_value)
loop_weight[row[mask]] = masked_weight
edge_index, _ = add_self_loops(edge_index)
edge_weight = torch.cat([edge_weight, loop_weight], dim=0)
```

**Comment by rusty1s** [comment]
It is maybe best to integrate `remove` and `add` into a single function for this kinds of operations. Will work on this.

**Comment by RuihongQiu** [comment]
Your solution is better than mine. I have just figured out using the positive and negative mask in `remove_self_loop` to record what is removed and add them back. It is a little bit unclean. Wait for your new update. : )

**Comment by RuihongQiu** [comment]
```python
def add_self_loops_partial(edge_index, edge_weight=None, fill_value=1, num_nodes=None):
    num_nodes = maybe_num_nodes(edge_index, num_nodes)
    row, col = edge_index
    mask = row == col
    masked_weight = edge_weight[mask]
    edge_index, edge_weight = remove_self_loops(edge_index, edge_weight)
    loop_weight = torch.full((num_nodes,), fill_value).cuda()
    loop_weight[row[mask]] = masked_weight
    edge_index, _ = add_self_loops(edge_index, num_nodes=num_nodes)
    edge_weight = torch.cat([edge_weight, loop_weight], dim=0)
    assert edge_index.shape[-1] == edge_weight.shape[0]
    return edge_index, edge_weight
```

Here is a runable solution based on your suggestion.

**Comment by rusty1s** [comment]
Thank you. Added to master.
